core/src/main/java/org/elasticsearch/cluster/routing/allocation/UnassignedShardDecision.java

```diff
@@ -0,0 +1,188 @@
+     */
+    public static UnassignedShardDecision noDecision(AllocationStatus allocationStatus,
+                                                     String explanation,
+                                                     Map<String, Decision> nodeDecisions) {
+        Objects.requireNonNull(explanation, "explanation must not be null");
+        Objects.requireNonNull(allocationStatus, "allocationStatus must not be null");
+        Objects.requireNonNull(nodeDecisions, "nodeDecisions must not be null");
+        return new UnassignedShardDecision(true, Decision.NO, allocationStatus, explanation, null, null, nodeDecisions);
+    }
+
+    /**
+     * Creates a THROTTLE decision with the given explanation and individual node-level decisions that
+     * comprised the final THROTTLE decision.
+     */
+    public static UnassignedShardDecision throttleDecision(String explanation,
+                                                           Map<String, Decision> nodeDecisions) {
+        Objects.requireNonNull(explanation, "explanation must not be null");
+        Objects.requireNonNull(nodeDecisions, "nodeDecisions must not be null");
+        return new UnassignedShardDecision(true, Decision.THROTTLE, AllocationStatus.DECIDERS_THROTTLED, explanation, null, null,
+                                           nodeDecisions);
+    }
+
+    /**
+     * Creates a YES decision with the given explanation and individual node-level decisions that
+     * comprised the final YES decision, along with the node id to which the shard is assigned and
+     * the allocation id for the shard, if available.
+     */
+    public static UnassignedShardDecision yesDecision(String explanation,
+                                                      String assignedNodeId,
+                                                      String allocationId,
+                                                      Map<String, Decision> nodeDecisions) {
+        Objects.requireNonNull(explanation, "explanation must not be null");
+        Objects.requireNonNull(assignedNodeId, "assignedNodeId must not be null");
+        Objects.requireNonNull(nodeDecisions, "nodeDecisions must not be null");
+        return new UnassignedShardDecision(true, Decision.YES, null, explanation, assignedNodeId, allocationId, nodeDecisions);
+    }
+
+    /**
+     * Returns <code>true</code> if a decision was taken by the allocator, {@code false} otherwise.
+     * If no decision was taken, then the rest of the fields in this object are meaningless and return {@code null}.
+     */
+    public boolean isDecisionTaken() {
+        return decisionTaken;
+    }
+
+    /**
+     * Returns the final decision made by the allocator on whether to assign the unassigned shard.
+     * This value can only be {@code null} if {@link #isDecisionTaken()} returns {@code false}.
+     */
+    @Nullable
+    public Decision getDecision() {
```

Comment:
> Additionally, I like calling these getFinalDecision in ClusterAllocationExplanation.java because it differentiated it from the node decisions, how do you feel about that? (It would also probably change getExplanation() to getFinalExplanation())

I think that's possible but would change how we handle decisions that never went through deciders.  For example, when handling the case where we are still pending the shard data fetch, we need an authoritative decision separate from what the node deciders decided (b/c in this case, we never went through the deciders), in order for the allocators to act on the decision.  Likewise with the explanations, we could lose context to create an appropriate explanation if we wait to do it outside in the `ClusterAllocationExplanation`, unless any given set of unique values for the `UnassignedShardDecision` will uniquely determine what the explanation should be.  Also, since we don't make a decision more than one time in any round on a given shard (i.e. only one of the allocators will make a decision on it, otherwise let the downstream allocator decide on it), its sort of easy to pick up the final decision and explanation as we process them.

What scenarios were you thinking of that would make this easier?